### ISSUE TITLE
Performance: quantity comparison to 0 is incredibly expensive

### ISSUE THREAD

**Issue by MaxNoe** [astropy/astropy#10396]
Quantities with units can be compared to 0 without a unit, in accordance with the interpretation that quantities are products of values and units.

However, this operation is forbiddingly expensive compared to the simple accessing of the value property:

```
In [1]: import astropy.units as u                                                                         

In [2]: q = 5 * u.m                                                                                       

In [3]: %timeit q != 0                                                                                    
42 µs ± 1.17 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [4]: %timeit q.value != 0                                                                              
607 ns ± 28 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

**Comment by dhomeier** [comment]
What exactly is "forbiddingly expensive"? It is much costlier, agreed, but comparison to another quantity takes almost as long (half the time on my system). So this seems to be mostly the overhead that units just bring about (and I suspect there could be some caching effects speeding up the `q.value` making the difference appear larger).
This does not mean there might not be some optimisation potential. Especially `q != 0` taking twice the time of `q != Q2` indicates there is some assignment or unassignment of units going on that could perhaps be saved with a smarter ordering of things.

**Comment by mhvk** [comment]
@MaxNoe - funny how something I really had to argue to implement as it was seen as too obscure and arguably not exact enough is now something obvious one wants faster! The problem essentially is that this is done as an add-on in the path of interacting between non-quantity and quantity, for the case regular unit conversion to dimensionless fails, in the `except` path:
https://github.com/astropy/astropy/blob/0a46d333a976d3d69c6cd7f9f2e3574298fe5209/astropy/units/quantity_helper/helpers.py#L64-L74

And then higher up that special-case is looked at in more detail again.

It may well be possible to speed it up, but so far all the effort has gone into speeding up the more common case, where one is comparing a dimensionless quantity with a plain array or number.

More generally, quantities are still quite slow compared to just working with numpy arrays, in part simply because of the overhead of unit conversion, and this shows up especially badly for scalars. I have on and off tried profiling to identify bottlenecks, and solved some of those (and also made some regressions were it really simplified the code...), but it may well be time for another look.

p.s. All this reminds me yet again that it would be really great if we not only had good documentation for users (which I think we can be proud of) but also for developers, about implementation and the logic behind it. But I never got beyond raising an issue for it... #8930

**Comment by mhvk** [comment]
p.s. Possibly should have clarified that the above happens in a place where unit conversion functions are created. At this place, only the function and the units are available, not the actual values. So, the `False` is used as a signals to actually check the values; this happens in
https://github.com/astropy/astropy/blob/0a46d333a976d3d69c6cd7f9f2e3574298fe5209/astropy/units/quantity_helper/converters.py#L168-L184

Basically, the logic is
1. One has no unit, try to convert quantity to dimensionless
2. If that fails, signal up to look at value
3. Check whether non-quantity is all zero or inf.

So, time comes both from the expensive unit conversion and a fairly expensive check that the non-quantity is all 0 or infinity.
